File: android/app/src/main/java/com/vortex/a3/core/earbuds/AudioOp.kt
package com.vortex.a3.core.earbuds

import org.json.JSONException
import org.json.JSONObject

sealed class AudioOp {
    object Request : AudioOp()
    object Claim : AudioOp()
    object Approve : AudioOp()
    data class Reject(val reason: RejectReason) : AudioOp()
    object Released : AudioOp()
    object Done : AudioOp()
    data class Failed(val stage: Stage, val message: String) : AudioOp()
}

enum class RejectReason(val wire: String) {
    Busy("busy"),
    InCall("in_call"),
    RecentSwitch("recent_switch"),
    Unknown("unknown");

    companion object {
        fun fromWire(s: String?): RejectReason? = entries.firstOrNull { it.wire == s }
    }
}

enum class Stage(val wire: String) {
    Disconnect("disconnect"),
    Connect("connect"),
    WaitReady("wait_ready");

    companion object {
        fun fromWire(s: String?): Stage? = entries.firstOrNull { it.wire == s }
    }
}

data class AudioOpFrame(
    val nonce: Long,
    val op: AudioOp,
    val mac: String,
    val ts: Long,
) {
    fun toJsonBytes(): ByteArray {
        val obj = JSONObject()
        obj.put("nonce", nonce)
        obj.put("op", encodeOp(op))
        obj.put("mac", mac)
        obj.put("ts", ts)
        return obj.toString().toByteArray(Charsets.UTF_8)
    }

    companion object {
        fun fromJsonBytes(bytes: ByteArray): AudioOpFrame? {
            return try {
                val obj = JSONObject(String(bytes, Charsets.UTF_8))
                val nonce = obj.optLong("nonce", -1L)
                if (nonce < 0) return null
                val opObj = obj.optJSONObject("op") ?: return null
                val op = decodeOp(opObj) ?: return null
                val mac = obj.optString("mac", "")
                val ts = obj.optLong("ts", 0L)
                if (mac.isBlank()) return null
                AudioOpFrame(nonce, op, mac, ts)
            } catch (_: JSONException) {
                null
            }
        }

        private fun encodeOp(op: AudioOp): JSONObject {
            val o = JSONObject()
            when (op) {
                AudioOp.Request -> o.put("kind", "request")
                AudioOp.Claim -> o.put("kind", "claim")
                AudioOp.Approve -> o.put("kind", "approve")
                is AudioOp.Reject -> {
                    o.put("kind", "reject")
                    o.put("reason", op.reason.wire)
                }
                AudioOp.Released -> o.put("kind", "released")
                AudioOp.Done -> o.put("kind", "done")
                is AudioOp.Failed -> {
                    o.put("kind", "failed")
                    o.put("stage", op.stage.wire)
                    o.put("message", op.message)
                }
            }
            return o
        }

        private fun decodeOp(o: JSONObject): AudioOp? {
            return when (o.optString("kind")) {
                "request" -> AudioOp.Request
                "claim" -> AudioOp.Claim
                "approve" -> AudioOp.Approve
                "reject" -> {
                    val r = RejectReason.fromWire(o.optString("reason")) ?: return null
                    AudioOp.Reject(r)
                }
                "released" -> AudioOp.Released
                "done" -> AudioOp.Done
                "failed" -> {
                    val s = Stage.fromWire(o.optString("stage")) ?: return null
                    val m = o.optString("message", "")
                    AudioOp.Failed(s, m)
                }
                else -> null
            }
        }
    }
}

File: linux/daemon/src/core/audio_op.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum AudioOp {
    Request,
    Claim,
    Approve,
    Reject { reason: RejectReason },
    Released,
    Done,
    Failed { stage: Stage, message: String },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RejectReason {
    Busy,
    InCall,
    RecentSwitch,
    Unknown,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Stage {
    Disconnect,
    Connect,
    WaitReady,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AudioOpFrame {
    pub nonce: u64,
    pub op: AudioOp,
    pub mac: String,
    pub ts: u64,
}

impl AudioOpFrame {
    pub fn to_json(&self) -> Result<Vec<u8>, serde_json::Error> {
        serde_json::to_vec(self)
    }

    pub fn from_json(bytes: &[u8]) -> Result<Self, serde_json::Error> {
        serde_json::from_slice(bytes)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn roundtrip_request() {
        let f = AudioOpFrame {
            nonce: 42,
            op: AudioOp::Request,
            mac: "AC:47:1B:25:71:C2".to_string(),
            ts: 1_700_000_000,
        };
        let bytes = f.to_json().unwrap();
        let back = AudioOpFrame::from_json(&bytes).unwrap();
        assert_eq!(back.nonce, 42);
        assert!(matches!(back.op, AudioOp::Request));
        assert_eq!(back.mac, f.mac);
    }

    #[test]
    fn roundtrip_reject_with_reason() {
        let f = AudioOpFrame {
            nonce: 100,
            op: AudioOp::Reject { reason: RejectReason::InCall },
            mac: "AA:BB:CC:DD:EE:FF".to_string(),
            ts: 0,
        };
        let bytes = f.to_json().unwrap();
        let back = AudioOpFrame::from_json(&bytes).unwrap();
        match back.op {
            AudioOp::Reject { reason } => assert_eq!(reason, RejectReason::InCall),
            _ => panic!("expected Reject"),
        }
    }

    #[test]
    fn roundtrip_failed() {
        let f = AudioOpFrame {
            nonce: 7,
            op: AudioOp::Failed { stage: Stage::Connect, message: "A2DP profile refused".into() },
            mac: "AA:BB:CC:DD:EE:FF".to_string(),
            ts: 1,
        };
        let bytes = f.to_json().unwrap();
        let back = AudioOpFrame::from_json(&bytes).unwrap();
        match back.op {
            AudioOp::Failed { stage, ref message } => {
                assert_eq!(stage, Stage::Connect);
                assert_eq!(message, "A2DP profile refused");
            }
            _ => panic!("expected Failed"),
        }
    }

    #[test]
    fn unknown_reject_reason_deserializes_to_unknown_or_error() {
        let bad =
            br#"{"nonce":1,"op":{"kind":"reject","reason":"future_reason"},"mac":"x","ts":0}"#;
        let r = AudioOpFrame::from_json(bad);
        assert!(r.is_err(), "unknown reason should error in V1");
    }
}
